tenacity/marshaller.py

```diff
@@ -0,0 +1,190 @@
+
+    def __str__(self):
+        return "RetryError[{0}]".format(self.last_attempt)
+
+
+class TryAgain(Exception):
+    """Always retry the executed function when raised."""
+
+
+class Future(futures.Future):
+    """Encapsulates a (future or past) attempted call to a target function."""
+
+    def __init__(self, attempt_number):
+        super(Future, self).__init__()
+        self.attempt_number = attempt_number
+
+    @property
+    def failed(self):
+        """Return whether a exception is being held in this future."""
+        return self.exception() is not None
+
+    @classmethod
+    def construct(cls, attempt_number, value, has_exception):
+        """Helper (for testing) for making these objects easily."""
+        fut = cls(attempt_number)
+        if has_exception:
+            fut.set_exception(value)
+        else:
+            fut.set_result(value)
+        return fut
+
+
+class Marshaller(object):
+    _local = threading.local()
+
+    def __init__(self,
+                 stop=stop_never, wait=wait_none(),
+                 retry=retry_if_exception_type(),
+                 before=before_nothing,
+                 after=after_nothing,
+                 reraise=False):
+        self._local.statistics = {}
+        self.stop = stop
+        self.wait = wait
+        self.retry = retry
+        self.before = before
+        self.after = after
+        self.reraise = reraise
+
+    @property
+    def statistics(self):
+        """A dictionary of runtime statistics this controller has gathered.
+
+        This dictionary will be empty when the controller has never been
+        ran. When it is running or has ran previously it should have (but
+        may not) have useful and/or informational keys and values when
+        running is underway and/or completed.
+
+        .. warning:: The keys in this dictionary **should** be some what
+                     stable (not changing), but there existence **may**
+                     change between major releases as new statistics are
+                     gathered or removed so before accessing keys ensure that
+                     they actually exist and handle when they do not.
+
+        .. note:: The values in this dictionary are local to the thread
+                  running call (so if multiple threads share the same retrying
+                  object - either directly or indirectly) they will each have
+                  there own view of statistics they have collected (in the
+                  future we may provide a way to aggregate the various
+                  statistics from each thread).
+        """
+
+        return self._local.statistics
+
+    def __repr__(self):
+        attrs = _utils.visible_attrs(self, attrs={'me': id(self)})
+        return ("<Marshaller object at 0x%(me)x (stop=%(stop)s, "
+                "wait=%(wait)s, sleep=%(sleep)s, retry=%(retry)s, "
+                "before=%(before)s, after=%(after)s)>") % (attrs)
+
+    def begin(self, fn):
```

Comment:
I changed Marshaller to be instanciated per decorated function rather than per call. This is why i need to split the decorate-init (`__init__`) and the call-init (`begin`).g